src/search/hnsw_indexer.cc

```diff
@@ -0,0 +1,514 @@
+
+    auto default_entry_node = GET_OR_RET(DefaultEntryPoint(level));
+    std::vector<NodeKey> entry_points{default_entry_node};
+
+    for (; level > target_level; level--) {
+      nearest_vec_items = GET_OR_RET(SearchLayer(level, inserted_vector_item, metadata->ef_runtime, entry_points));
+      entry_points = {nearest_vec_items[0].key};
+    }
+
+    for (; level >= 0; level--) {
+      nearest_vec_items = GET_OR_RET(SearchLayer(level, inserted_vector_item, metadata->ef_construction, entry_points));
+      auto candidate_vec_items = GET_OR_RET(SelectNeighbors(inserted_vector_item, nearest_vec_items, level));
+      auto node = Node(std::string(key), level);
+      auto m_max = level == 0 ? 2 * metadata->m : metadata->m;
+
+      std::unordered_set<NodeKey> connected_edges_set;
+      std::unordered_map<NodeKey, std::unordered_set<NodeKey>> deleted_edges_map;
+
+      // Check if candidate node has room for more outgoing edges
+      auto has_room_for_more_edges = [&](uint16_t candidate_node_num_neighbours) {
+        return candidate_node_num_neighbours < m_max;
+      };
+
+      // Check if candidate node has room after some other nodes' are pruned in current batch
+      auto has_room_after_deletions = [&](const Node& candidate_node, uint16_t candidate_node_num_neighbours) {
+        auto it = deleted_edges_map.find(candidate_node.key);
+        if (it != deleted_edges_map.end()) {
+          auto num_deleted_edges = static_cast<uint16_t>(it->second.size());
+          return (candidate_node_num_neighbours - num_deleted_edges) < m_max;
+        }
+        return false;
+      };
+
+      for (const auto& candidate_vec : candidate_vec_items) {
+        auto candidate_node = Node(candidate_vec.key, level);
+        auto candidate_node_metadata = GET_OR_RET(candidate_node.DecodeMetadata(search_key, storage));
+        uint16_t candidate_node_num_neighbours = candidate_node_metadata.num_neighbours;
+
+        if (has_room_for_more_edges(candidate_node_num_neighbours) ||
+            has_room_after_deletions(candidate_node, candidate_node_num_neighbours)) {
+          GET_OR_RET(AddEdge(inserted_vector_item.key, candidate_node.key, level, batch));
+          connected_edges_set.insert(candidate_node.key);
+          continue;
+        }
+
+        // Re-evaluate the neighbours for the candidate node
+        candidate_node.DecodeNeighbours(search_key, storage);
+        auto candidate_node_neighbour_vec_items =
+            GET_OR_RET(DecodeNodesToVectorItems(candidate_node.neighbours, level, search_key, storage, metadata));
+        candidate_node_neighbour_vec_items.push_back(inserted_vector_item);
+        auto sorted_neighbours = GET_OR_RET(SelectNeighbors(candidate_vec, candidate_node_neighbour_vec_items, level));
+
+        bool inserted_node_is_selected = std::find(sorted_neighbours.begin(), sorted_neighbours.end(),
+                                                   inserted_vector_item) != sorted_neighbours.end();
```

Comment:
If it's sorted maybe we can use `std::binary_search` here.

Reply:
`SelectNeighbors` returns `sorted_neighbours` (`std::vector<VectorItem>`) based on distance, while `VectorItem` overloads `operator<` based on its `key`. So I guess `std::binary_search` wouldn't work here since `sorted_neighbours` is sorted by distance instead of keys. 

I updated `sorted_neighbours` into `sorted_neighbours_by_distance` to make it more clear. Let me know if I misunderstand anything here :)